mockrxandroidble/src/main/java/com/polidea/rxandroidble/mockrxandroidble/RxBleClientMock.java

```diff
@@ -0,0 +1,285 @@
+         * Should be used in pair with {@link CharacteristicsBuilder#addCharacteristic}
+         */
+        public DescriptorsBuilder() {
+            this.bluetoothGattDescriptors = new ArrayList<>();
+        }
+
+        /**
+         * Adds a {@link BluetoothGattDescriptor} with specified parameters.
+         *
+         * @param uuid descriptor UUID
+         * @param data locally stored value of the descriptor
+         */
+        public DescriptorsBuilder addDescriptor(@NonNull UUID uuid, @NonNull byte[] data) {
+            BluetoothGattDescriptor bluetoothGattDescriptor = new BluetoothGattDescriptor(uuid, 0);
+            bluetoothGattDescriptor.setValue(data);
+            bluetoothGattDescriptors.add(bluetoothGattDescriptor);
+            return this;
+        }
+
+        /**
+         * Create the  {@link List} of {@link BluetoothGattDescriptor} using the configured values.
+         */
+        public List<BluetoothGattDescriptor> build() {
+            return bluetoothGattDescriptors;
+        }
+    }
+
+    private final Map<String, RxBleDevice> discoverableDevices;
+
+    private RxBleClientMock(Builder builder) {
+        discoverableDevices = builder.discoverableDevices;
+    }
+
+    @Override
+    public RxBleDevice getBleDevice(@NonNull String macAddress) {
+        RxBleDevice rxBleDevice = discoverableDevices.get(macAddress);
+
+        if (rxBleDevice == null) {
+            throw new IllegalStateException("Mock is not configured for a given mac address. Use Builder#addDevice method.");
+        }
+
+        return rxBleDevice;
+    }
+
+    @Override
+    public Observable<RxBleScanResult> scanBleDevices(@Nullable UUID... filterServiceUUIDs) {
+        return createScanOperation(filterServiceUUIDs);
+    }
+
+    private RxBleScanResult convertToPublicScanResult(RxBleDevice bleDevice, Integer rssi, byte[] scanRecord) {
+        return new RxBleScanResult(bleDevice, rssi, scanRecord);
+    }
+
+    @NonNull
+    private Observable<RxBleScanResult> createScanOperation(@Nullable UUID[] filterServiceUUIDs) {
+        return Observable.defer(() -> Observable.from(discoverableDevices.values())
+                .filter(rxBleDevice -> filterDevice(rxBleDevice, filterServiceUUIDs))
+                .map(rxBleDevice -> {
+                    RxBleDeviceMock rxBleDeviceMock = (RxBleDeviceMock) rxBleDevice;
+                    return convertToPublicScanResult(rxBleDeviceMock, rxBleDeviceMock.getRssi(), rxBleDeviceMock.getScanRecord());
+                })
+                .share());
+    }
+
+    private boolean filterDevice(RxBleDevice rxBleDevice, @Nullable UUID[] filterServiceUUIDs) {
+
+        if (filterServiceUUIDs == null || filterServiceUUIDs.length == 0) {
+            return true;
+        }
+
+        List<UUID> advertisedUUIDs = ((RxBleDeviceMock) rxBleDevice).getAdvertisedUUIDs();
+
+        for (UUID desiredUUID : filterServiceUUIDs) {
+
+            if (advertisedUUIDs.contains(desiredUUID)) {
```

Comment:
This is wrong, current implementation requires all filterServiceUUIDs to be contained by the advertised device not just one.

`
private boolean containsDesiredServiceIds(byte[] scanRecord) {
        List<UUID> advertisedUUIDs = uuidUtil.extractUUIDs(scanRecord);

        for (UUID desiredUUID : filterServiceUUIDs) {

            if (!advertisedUUIDs.contains(desiredUUID)) {
                return false;
            }
        }

        return true;
    }
`